File: android/src/main/java/com/zaions/capacitor/firebasekit/services/PerformanceService.kt
package com.zaions.capacitor.firebasekit.services

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.google.firebase.Firebase
import com.google.firebase.perf.FirebasePerformance
import com.google.firebase.perf.metrics.Trace
import com.google.firebase.perf.performance
import java.util.concurrent.ConcurrentHashMap

class PerformanceService(private val plugin: Plugin) {
    
    private val performance: FirebasePerformance by lazy {
        Firebase.performance
    }
    
    private val traces = ConcurrentHashMap<String, Trace>()
    
    fun initialize(call: PluginCall) {
        val enabled = call.getBoolean("enabled", true) ?: true
        performance.isPerformanceCollectionEnabled = enabled
        call.resolve()
    }
    
    fun setPerformanceCollectionEnabled(call: PluginCall) {
        val enabled = call.getBoolean("enabled") ?: return
        performance.isPerformanceCollectionEnabled = enabled
        call.resolve()
    }
    
    fun isPerformanceCollectionEnabled(call: PluginCall) {
        val result = JSObject()
        result.put("enabled", performance.isPerformanceCollectionEnabled)
        call.resolve(result)
    }
    
    fun startTrace(call: PluginCall) {
        val traceName = call.getString("traceName") ?: return
        val traceId = "trace_${System.currentTimeMillis()}_${traceName.hashCode()}"
        
        val trace = performance.newTrace(traceName)
        trace.start()
        traces[traceId] = trace
        
        val result = JSObject()
        result.put("traceId", traceId)
        call.resolve(result)
    }
    
    fun stopTrace(call: PluginCall) {
        val traceId = call.getString("traceId") ?: return
        val trace = traces.remove(traceId)
        
        if (trace != null) {
            trace.stop()
            call.resolve()
        } else {
            call.reject("Trace not found: $traceId")
        }
    }
    
    fun incrementMetric(call: PluginCall) {
        val traceId = call.getString("traceId") ?: return
        val metricName = call.getString("metricName") ?: return
        val value = call.getInt("value") ?: 1
        
        val trace = traces[traceId]
        if (trace != null) {
            trace.incrementMetric(metricName, value.toLong())
            call.resolve()
        } else {
            call.reject("Trace not found: $traceId")
        }
    }
    
    fun setMetric(call: PluginCall) {
        val traceId = call.getString("traceId") ?: return
        val metricName = call.getString("metricName") ?: return
        val value = call.getInt("value") ?: return
        
        val trace = traces[traceId]
        if (trace != null) {
            trace.putMetric(metricName, value.toLong())
            call.resolve()
        } else {
            call.reject("Trace not found: $traceId")
        }
    }
    
    fun getMetric(call: PluginCall) {
        val traceId = call.getString("traceId") ?: return
        val metricName = call.getString("metricName") ?: return
        
        val trace = traces[traceId]
        if (trace != null) {
            val value = trace.getLongMetric(metricName)
            val result = JSObject()
            result.put("value", value)
            call.resolve(result)
        } else {
            call.reject("Trace not found: $traceId")
        }
    }
    
    fun putAttribute(call: PluginCall) {
        val traceId = call.getString("traceId") ?: return
        val attribute = call.getString("attribute") ?: return
        val value = call.getString("value") ?: return
        
        val trace = traces[traceId]
        if (trace != null) {
            trace.putAttribute(attribute, value)
            call.resolve()
        } else {
            call.reject("Trace not found: $traceId")
        }
    }
    
    fun getAttributes(call: PluginCall) {
        val traceId = call.getString("traceId") ?: return
        
        val trace = traces[traceId]
        if (trace != null) {
            val attributes = trace.attributes
            val result = JSObject()
            result.put("attributes", JSObject().apply {
                attributes.forEach { (key, value) ->
                    put(key, value)
                }
            })
            call.resolve(result)
        } else {
            call.reject("Trace not found: $traceId")
        }
    }
    
    fun removeAttribute(call: PluginCall) {
        val traceId = call.getString("traceId") ?: return
        val attribute = call.getString("attribute") ?: return
        
        val trace = traces[traceId]
        if (trace != null) {
            trace.removeAttribute(attribute)
            call.resolve()
        } else {
            call.reject("Trace not found: $traceId")
        }
    }
    
    fun startScreenTrace(call: PluginCall) {
        val screenName = call.getString("screenName") ?: return
        val traceId = "screen_${System.currentTimeMillis()}_${screenName.hashCode()}"
        
        val trace = performance.newTrace("_st_$screenName")
        trace.start()
        traces[traceId] = trace
        
        val result = JSObject()
        result.put("traceId", traceId)
        call.resolve(result)
    }
    
    fun stopScreenTrace(call: PluginCall) {
        stopTrace(call)
    }
    
    fun recordNetworkRequest(call: PluginCall) {
        // TODO: Implement network request recording
        call.resolve()
    }
}

File: ios/Plugin/Services/PerformanceService.swift
import Foundation
import Capacitor
import FirebasePerformance

class PerformanceService {
    private weak var plugin: CAPPlugin?
    private var traces: [String: Trace] = [:]
    
    init(plugin: CAPPlugin) {
        self.plugin = plugin
    }
    
    func initialize(_ call: CAPPluginCall) {
        let enabled = call.getBool("enabled", true)
        Performance.sharedInstance().isDataCollectionEnabled = enabled
        call.resolve()
    }
    
    func setPerformanceCollectionEnabled(_ call: CAPPluginCall) {
        guard let enabled = call.getBool("enabled") else {
            call.reject("Enabled parameter is required")
            return
        }
        
        Performance.sharedInstance().isDataCollectionEnabled = enabled
        call.resolve()
    }
    
    func isPerformanceCollectionEnabled(_ call: CAPPluginCall) {
        let enabled = Performance.sharedInstance().isDataCollectionEnabled
        call.resolve(["enabled": enabled])
    }
    
    func startTrace(_ call: CAPPluginCall) {
        guard let traceName = call.getString("traceName") else {
            call.reject("Trace name is required")
            return
        }
        
        guard let trace = Performance.startTrace(name: traceName) else {
            call.reject("Failed to start trace")
            return
        }
        
        let traceId = "trace_\(Date().timeIntervalSince1970)_\(traceName.hashValue)"
        traces[traceId] = trace
        
        call.resolve(["traceId": traceId])
    }
    
    func stopTrace(_ call: CAPPluginCall) {
        guard let traceId = call.getString("traceId") else {
            call.reject("Trace ID is required")
            return
        }
        
        guard let trace = traces.removeValue(forKey: traceId) else {
            call.reject("Trace not found: \(traceId)")
            return
        }
        
        trace.stop()
        call.resolve()
    }
    
    func incrementMetric(_ call: CAPPluginCall) {
        guard let traceId = call.getString("traceId"),
              let metricName = call.getString("metricName") else {
            call.reject("Trace ID and metric name are required")
            return
        }
        
        let value = call.getInt("value", 1)
        
        guard let trace = traces[traceId] else {
            call.reject("Trace not found: \(traceId)")
            return
        }
        
        trace.incrementMetric(metricName, by: Int64(value))
        call.resolve()
    }
    
    func setMetric(_ call: CAPPluginCall) {
        guard let traceId = call.getString("traceId"),
              let metricName = call.getString("metricName"),
              let value = call.getInt("value") else {
            call.reject("Trace ID, metric name, and value are required")
            return
        }
        
        guard let trace = traces[traceId] else {
            call.reject("Trace not found: \(traceId)")
            return
        }
        
        trace.setIntValue(Int64(value), forMetric: metricName)
        call.resolve()
    }
    
    func getMetric(_ call: CAPPluginCall) {
        guard let traceId = call.getString("traceId"),
              let metricName = call.getString("metricName") else {
            call.reject("Trace ID and metric name are required")
            return
        }
        
        guard let trace = traces[traceId] else {
            call.reject("Trace not found: \(traceId)")
            return
        }
        
        let value = trace.valueForIntMetric(metricName)
        call.resolve(["value": value])
    }
    
    func putAttribute(_ call: CAPPluginCall) {
        guard let traceId = call.getString("traceId"),
              let attribute = call.getString("attribute"),
              let value = call.getString("value") else {
            call.reject("Trace ID, attribute, and value are required")
            return
        }
        
        guard let trace = traces[traceId] else {
            call.reject("Trace not found: \(traceId)")
            return
        }
        
        trace.setValue(value, forAttribute: attribute)
        call.resolve()
    }
    
    func getAttributes(_ call: CAPPluginCall) {
        guard let traceId = call.getString("traceId") else {
            call.reject("Trace ID is required")
            return
        }
        
        guard let trace = traces[traceId] else {
            call.reject("Trace not found: \(traceId)")
            return
        }
        
        call.resolve(["attributes": trace.attributes])
    }
    
    func removeAttribute(_ call: CAPPluginCall) {
        guard let traceId = call.getString("traceId"),
              let attribute = call.getString("attribute") else {
            call.reject("Trace ID and attribute are required")
            return
        }
        
        guard let trace = traces[traceId] else {
            call.reject("Trace not found: \(traceId)")
            return
        }
        
        trace.removeAttribute(attribute)
        call.resolve()
    }
    
    func startScreenTrace(_ call: CAPPluginCall) {
        guard let screenName = call.getString("screenName") else {
            call.reject("Screen name is required")
            return
        }
        
        let traceName = "_st_\(screenName)"
        guard let trace = Performance.startTrace(name: traceName) else {
            call.reject("Failed to start screen trace")
            return
        }
        
        let traceId = "screen_\(Date().timeIntervalSince1970)_\(screenName.hashValue)"
        traces[traceId] = trace
        
        call.resolve(["traceId": traceId])
    }
    
    func stopScreenTrace(_ call: CAPPluginCall) {
        stopTrace(call)
    }
    
    func recordNetworkRequest(_ call: CAPPluginCall) {
        // TODO: Implement network request recording
        call.resolve()
    }
}